### ISSUE TITLE
Starting WSL2 from a windows service - crash

### ISSUE THREAD

**Issue by redpin1441** [microsoft/WSL2-Linux-Kernel#125]
Running win10 pro Insider Preview
Ubuntu 18.04
OS Build 20150.1000

I've written a c# application that starts wsl like this (using Process class)
c:\\windows\\sysnative\\wsl.exe -u myuser-- ~/start.sh

When invoking the app from a cmd prompt, the default wsl instance runs fine.

However, if whilst logged in, I manually start the same process as a service, I have found that wsl -l -v does not report the state as running. 
Further investigation show that there is a crash which is 100% reproducable. It appears there is a possible STATUS_STACK_BUFFER_OVERRUN. In the event log, I see these 2 events

Faulting application name: wsl.exe, version: 10.0.20150.1000, time stamp: 0x4037abbd
Faulting module name: ucrtbase.dll, version: 10.0.20150.1000, time stamp: 0x66e3e5e0
Exception code: 0xc0000409
Fault offset: 0x000000000007b05e
Faulting process id: 0x1594
Faulting application start time: 0x01d6461bb36b8625
Faulting application path: C:\windows\system32\wsl.exe
Faulting module path: C:\WINDOWS\System32\ucrtbase.dll
Report Id: cee38d9c-5e96-47e8-8752-c2ece1da8523
Faulting package full name: 
Faulting package-relative application ID: 

This is followed by this event;

Fault bucket 1798145230619860747, type 5
Event Name: BEX64
Response: Not available
Cab Id: 0

Problem signature:
P1: wsl.exe
P2: 10.0.20150.1000
P3: 4037abbd
P4: ucrtbase.dll
P5: 10.0.20150.1000
P6: 66e3e5e0
P7: 000000000007b05e
P8: c0000409
P9: 0000000000000007
P10: 

Attached files:
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER7997.tmp.mdmp
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER79D6.tmp.WERInternalMetadata.xml
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER7A35.tmp.xml
WPR_initiated_DiagTrackMiniLogger_OneTrace User Logger 20200512 1 Event Collector_0_inject.etl
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER7BDC.tmp.etl
WPR_initiated_DiagTrackMiniLogger_WPR System Collector_inject.etl
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER7BDD.tmp.etl
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER7BE8.tmp.csv
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER7C09.tmp.txt

These files may be available here:
\\?\C:\ProgramData\Microsoft\Windows\WER\ReportArchive\AppCrash_wsl.exe_3b6dbfe7a32d540c82c9bc14343adfa5c02fb6_96f3f2d4_7db15722-ee27-4fe3-9c84-7e3e76063703

Analysis symbol: 
Rechecking for solution: 0
Report Id: cee38d9c-5e96-47e8-8752-c2ece1da8523
Report Status: 268435456
Hashed bucket: 8211a63e9204e54078f44b8f6da8df0b

It is possible to start wsl from a service? 

Are any of the debug files required?

Regards,
Jason.

**Comment by levicki** [comment]
Your exception code `0xC0000409` means `STATUS_STACK_BUFFER_OVERRUN`. I am guessing they will need to see your minidump file from that crash.

On a side note, starting a process (which is supposed to interact with user window session or at least with user environment) from a Windows Service is not just a matter of using Process class in C#. Services run in isolated session and under `NT AUTHORITY\SYSTEM` account which means that any processes you launch have no access to the Windows user profile and interactive window session.

Most likely you will need to [start a process as a Windows user](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessasusera) under which wsl distro was installed in order for this to work.

**Comment by redpin1441** [comment]
'On a side note, starting a process (which is supposed to interact with user window session or at least with user environment) from a Windows Service is not just a matter of using Process class in C#.'
This is a shame that it has to work in this way. You'd have thought the default kernel would be able to run irrespective of the user. The issue I have with this approach is that the service is likely to be installed on customer machines where the login user is unknown, so impersonating an unknown user could be tricky,. Having said that, its not the end of the world that it works this way as the customer is likely to be using a windows application on this PC anyway and therefore would be signed in. I just thought I'd try it and see what happens.

**Comment by levicki** [comment]
What is your need for a Windows service that starts WSL anyway? Can't you just run WSL as a startup task when a user logs in?

If you wanted to run it independent from any Windows user being logged in then frankly I am not sure whether that's currently possible. Microsoft developers would know the answer to that.

**Comment by levicki** [comment]
To clarify -- as far as I am aware, when you first launch a distro you installed from the Windows Store, you are ssked to create a WSL user which is then tied to a currently logged on Windows user account so that every time you start WSL it inherits your Windows account token for authentication and identification. That allows running bash and accessing your files without actually having to log in to WSL every time with the user you created.

You can test how WSL behaves when run under a diferent account using `psexec -s` command to run it under `NT AUTHORITY\SYSTEM` -- in essence the same thing you do when you launch it from a Windows service.

If you absolutely need to run it from the service and if you are sure that there will never be more than one Windows user logged on then you can get a user token for impersonation from `explorer.exe` process but I do not recommend going that route -- it is a hack and it will be prone to failure.

TL;DR -- simply placing a shortcut to run WSL into user's Start Menu Startup folder should work without needing Windows service with complex impersonation trickery.
